src/dodal/devices/areadetector/plugins/MJPG.py

```diff
@@ -1,138 +1,88 @@
+        self.KICKOFF_TIMEOUT = 30.0
+
+        super().__init__(name)
+
+    async def _save_image(self, image: Image.Image):
+        """A helper function to save a given image to the path supplied by the \
+            directory and filename signals. The full resultant path is put on the \
+            last_saved_path signal
         """
-        filename_str = self.filename.get()
-        directory_str: str = self.directory.get()  # type: ignore
+        filename_str = await self.filename.get_value()
+        directory_str = await self.directory.get_value()
 
         path = Path(f"{directory_str}/{filename_str}.png").as_posix()
-        if not os.path.isdir(Path(directory_str)):
+        if not Path(directory_str).is_dir():
             LOGGER.info(f"Snapshot folder {directory_str} does not exist, creating...")
-            os.mkdir(directory_str)
+            Path(directory_str).mkdir(parents=True)
 
         LOGGER.info(f"Saving image to {path}")
-        image.save(path)
-        self.last_saved_path.put(path)
 
-    def trigger(self):
+        buffer = BytesIO()
+        image.save(buffer, format="png")
+        async with aiofiles.open(path, "wb") as fh:
+            await fh.write(buffer.getbuffer())
+        await self.last_saved_path.set(path, wait=True)
+
+    @AsyncStatus.wrap
+    async def trigger(self):
         """This takes a snapshot image from the MJPG stream and send it to the
         post_processing method, expected to be implemented by a child of this class.
 
         It is the responsibility of the child class to save any resulting images.
         """
-        st = DeviceStatus(device=self, timeout=self.KICKOFF_TIMEOUT)
-        url_str = self.url.get()
-
-        assert isinstance(
-            self.oav_params, OAVConfigParams
-        ), "MJPG does not have valid OAV parameters"
-        self.microns_per_pixel_x.set(self.oav_params.micronsPerXPixel)
-        self.microns_per_pixel_y.set(self.oav_params.micronsPerYPixel)
-
-        def get_snapshot():
-            try:
-                response = requests.get(url_str, stream=True)
-                response.raise_for_status()
-                with Image.open(BytesIO(response.content)) as image:
-                    self.post_processing(image)
-                    st.set_finished()
-            except requests.HTTPError as e:
-                st.set_exception(e)
-
-        threading.Thread(target=get_snapshot, daemon=True).start()
-
-        return st
+        url_str = await self.url.get_value()
+
+        async with ClientSession() as session:
+            async with session.get(url_str) as response:
+                if not response.ok:
+                    LOGGER.error(
+                        f"OAV responded with {response.status}: {response.reason}."
+                    )
+                    raise ClientConnectionError(
+                        f"OAV responded with {response.status}: {response.reason}."
+                    )
```

Comment:
Should: I think we can just do `response.raise_for_status()` or actually `ClientSession(raise_for_status=True)`

Reply:
I tried using this `response.raise_for_status()` and couldn't make it work, all the tests kept failing because they couldn't get past it - even when mocking it. Haven't managed to figure out why yet but this at least did the job. Didn't know it could be passed as an argument to `ClientSession` so will try that